### DISCUSSION TITLE
Parser

### DISCUSSION THREAD

**Discussion by rsiddu0511** [AMReX-Codes/amrex#2130] [Q&A] [answered]
I used "Parser" function to find the value of variables using a set of constants.
But I encountered the following error:

`Parser is not a member of amrex`

I used the following lines in my GNUMakeFile.

`amrex::Parser parser("a=(b*c)/(d*d)");`

`parser.setConstant(b);`
`parser.setConstant(c);`
`parser.setConstant(d);`
`parser.registerVariables({"a"});`
`auto f = parser.compile<2>();`

Also, the compilation seems to have some issues too:
`expected primary-expression before ‘)’ token`

**Comment by WeiqunZhang** [comment, marked as answer]
> `Parser is not a member of amrex`

Is there `#include <AMReX_Parser.H>`?

> I used the following lines in my GNUMakeFile.

I assume this is a typo.  Parser is for C++ code, not for make.
 
> `amrex::Parser parser("a=(b*c)/(d*d)");`

This is not a valid code for the parser.  What do you want to compute? `(b*c)/(d*d)`? Assignment like `a=...;` can be used to define a local variable that's internal to the parser.  But the parser expression must have an expression as the last statement, and that will be its return value when `ParserExecutor<N>::operator()` is called.  `a=(b*c)/(d*d); a` is valid, but then `a` is not a variable to be registered with `registerVariables`.  Parser is usually used to evaluate a runtime string.  If it's a string literal known at compile time, one should just use a C++ function for it.
 
> `parser.setConstant(b);`

The function `void setConstant (std::string const& name, amrex::Real c);` takes two arguments.

> `parser.registerVariables({"a"});`
> `auto f = parser.compile<2>();`
